sycl/tools/lspi.cpp

```diff
@@ -0,0 +1,167 @@
+//==----------- lspi.cpp ---------------------------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// The "lspi" utility lists all platforms/devices discovered by PI similar to
+// how lscl prints this for OpenCL devices. It can probably be eventually merged
+// with the more complex "check-sycl" utility.
+//
+// There are two types of output:
+//   concise (default) and
+//   verbose (enabled with any argument).
+//
+// In verbose mode it also prints, which devices would be chose  by various SYCL
+// device selectors.
+//
+#include <CL/sycl.hpp>
+
+#include <cstdlib>
+#include <iostream>
+#include <stdlib.h>
+
+using namespace cl::sycl;
+using namespace std;
+
+// Controls verbose output vs. concise.
+bool verbose = false;
+
+// Trivial custom selector that selects a device of the given type.
+class custom_selector : public device_selector {
+  info::device_type MType;
+
+public:
+  custom_selector(info::device_type Type) : MType(Type) {}
+  int operator()(const device &Dev) const override {
+    return Dev.get_info<info::device::device_type>() == MType ? 1 : -1;
+  }
+};
+
+static void printDeviceInfo(const device &Device, const string &Prepend) {
+  auto DeviceType = Device.get_info<info::device::device_type>();
+  string DeviceTypeName;
+  switch (DeviceType) {
+  case info::device_type::cpu:
+    DeviceTypeName = "CPU ";
+    break;
+  case info::device_type::gpu:
+    DeviceTypeName = "GPU ";
+    break;
+  case info::device_type::host:
+    DeviceTypeName = "HOST";
+    break;
+  case info::device_type::accelerator:
+    DeviceTypeName = "ACC ";
+    break;
+  default:
+    DeviceTypeName = "??? ";
+    break;
+  }
+
+  auto DeviceVersion = Device.get_info<info::device::version>();
+  auto DeviceName = Device.get_info<info::device::name>();
+  auto DeviceVendor = Device.get_info<info::device::vendor>();
+  auto DeviceDriverVersion = Device.get_info<info::device::driver_version>();
+
+  if (verbose) {
+    cout << Prepend << "Type       : " << DeviceTypeName << std::endl;
```

Comment:
Seems weird to have cout without std but std::endl at the end

Reply:
the "endl" was conflicting with something in cl::sycl
removed using std now